lms/static/scripts/frontend_apps/components/SubmitGradeForm.tsx

```diff
@@ -99,14 +100,15 @@ export default function SubmitGradeForm({
   const [draftGradeValue, setDraftGradeValue] = useState<string | null>(null);
 
   // Track if current grade has changed compared to what was originally loaded
-  const [hasUnsavedChanges, setHasUnsavedChanges] = useState<boolean>(false);
-  const updateHasUnsavedChanges = useCallback(
-    (hasUnsavedChanges: boolean) => {
-      setHasUnsavedChanges(hasUnsavedChanges);
-      onUnsavedChanges?.(hasUnsavedChanges);
-    },
-    [onUnsavedChanges]
+  const hasUnsavedChanges = useMemo(
+    () => draftGradeValue !== null && draftGradeValue !== grade.data,
+    [draftGradeValue, grade.data]
   );
+  const prevHasUnsavedChanges = useRef(hasUnsavedChanges);
+  if (prevHasUnsavedChanges.current !== hasUnsavedChanges) {
+    onUnsavedChanges?.(hasUnsavedChanges);
+  }
+  prevHasUnsavedChanges.current = hasUnsavedChanges;
 
   // Make sure instructors are notified if there's a risk to lose unsaved data
   useWarnOnPageUnload(hasUnsavedChanges);
```

Comment:
There is an effect just below this code which resets the draft when the user changes. That is an instance where a state update on render should be used instead of an effect.

Reply:
Gotcha!